packages/web-components/src/calendar/calendar.ts

```diff
@@ -0,0 +1,687 @@
+    }-${this.navigatedDate.getDate()}-${this.navigatedDate.getFullYear()}`;
+
+    const el = this.shadowRoot?.querySelector(`slot[name=${CSS.escape(navigatedDateString)}]`)
+      ?.parentElement as HTMLElement;
+
+    return el;
+  }
+
+  /**
+   * Handles keyboard events on a secondary panel cell
+   * @param event - Keyboard event
+   * @param detail - the year or month of the event target cell
+   */
+  public handleSecondaryPanelKeydown(event: KeyboardEvent, detail: number): boolean {
+    const currentCell = event.target as HTMLElement;
+    currentCell.tabIndex = -1;
+
+    if (!this.secondaryPanelCells) {
+      return false;
+    }
+
+    // Get the index of the event target cell in the secondaryPanelCells array
+    let index = this.secondaryPanelCells.indexOf(currentCell);
+
+    switch (event.key) {
+      case keyEnter: {
+        this.$emit('secondaryPanelCellSelected', detail);
+        break;
+      }
+      case keyArrowRight: {
+        event.preventDefault();
+
+        // Check if reached the end of the current year/decade
+        if (index === this.secondaryPanelCells.length - 1) {
+          this.yearPickerOpen
+            ? (this.yearPickerDecade = this.getYearPickerInfo().nextStart)
+            : (this.monthPickerYear = this.getMonthPickerInfo().next);
+          return true;
+        }
+        index = (index + 1) % this.secondaryPanelCells.length;
+        break;
+      }
+      case keyArrowLeft: {
+        event.preventDefault();
+        if (index === 0) {
+          this.yearPickerOpen
+            ? (this.yearPickerDecade = this.getYearPickerInfo().previousStart)
+            : (this.monthPickerYear = this.getMonthPickerInfo().previous);
+          return true;
+        }
+        index = (index - 1) % this.secondaryPanelCells.length;
+        break;
+      }
+      case keyArrowDown: {
+        event.preventDefault();
+        if (index >= this.secondaryPanelCells.length - 4) {
+          this.yearPickerOpen
+            ? (this.yearPickerDecade = this.getYearPickerInfo().nextStart)
+            : (this.monthPickerYear = this.getMonthPickerInfo().next);
+        }
+        index = (index + 4) % this.secondaryPanelCells.length;
+        break;
+      }
+      case keyArrowUp: {
+        event.preventDefault();
+        if (index < 4) {
+          this.yearPickerOpen
+            ? (this.yearPickerDecade = this.getYearPickerInfo().previousStart)
+            : (this.monthPickerYear = this.getMonthPickerInfo().previous);
+        }
+        index = (index - 4) % this.secondaryPanelCells.length;
+        break;
+      }
+      case keyHome: {
+        event.preventDefault();
+        // Update focus to the first cell on the panel
+        index = 0;
+        break;
+      }
+      case keyEnd: {
+        event.preventDefault();
+
+        // Update focus to the last cell on the panel
+        index = this.secondaryPanelCells.length - 1;
+        break;
+      }
+      default:
+        break;
+    }
```

Comment:
I wonder if it makes sense to break this into smaller methods? This function is getting rather large and is doing a lot. You could create simple methods for what each keypress does, e.g.
```typescript
public handleSecondaryPanelKeydown(event: KeyboardEvent, detail: number): boolean {
  switch (event.key) {
    case keyEnter:
      this.handleEnterKey(detail);
      break;
    case keyArrowRight:
      this.handleRightArrowKey();
      break;
    case keyArrowLeft:
      this.handleLeftArrowKey();
      break;
    // other cases...
  }
  // rest of code...
}

private handleEnterKey(detail: number): void {
  this.$emit('secondaryPanelCellSelected', detail);
}

private handleRightArrowKey(): void {
  // Logic for handling right arrow key
}

private handleLeftArrowKey(): void {
  // Logic for handling left arrow key
}
```

Reply:
That would definitely make the keydown handler cleaner, but my concern is that each of those helper functions would look very similar to each other. Maybe I can refactor some of the repeating code into a helper method.

Reply:
Ultimately it'll be up to you. The only reason I mention doing it this way is for your future self ;) Looking back at this code in 6-12 months, will it be as clear to you as it is now?